src/observer.ts

```diff
@@ -12,50 +14,57 @@ export function observer<P>(baseComponent: FunctionComponent<P>): FunctionCompon
     if (isUsingStaticRendering) {
         return baseComponent
     }
+
     const baseComponentName = baseComponent.displayName || baseComponent.name
+
     // memo; we are not intested in deep updates
     // in props; we assume that if deep objects are changed,
     // this is in observables, which would have been tracked anyway
-
-    const memoComponent = memo(props => {
-        // forceUpdate 2.0
-        const forceUpdate = useForceUpdate()
-
-        // create a Reaction once, and memoize it
-        const reaction = useMemo(
-            () =>
-                // If the Reaction detects a change in dependency,
-                // force a new render
-                new Reaction(
-                    `observer(${baseComponentName})`,
-                    forceUpdate
-                ),
-            []
-        )
-
-        // clean up the reaction if this component is unMount
-        useUnmount(() => reaction.dispose())
+    const memoComponent = memo((props: P) => {
+        const observerReaction = useObserverReaction(baseComponentName)
 
         // render the original component, but have the
         // reaction track the observables, so that rendering
         // can be invalidated (see above) once a dependency changes
-        let rendering
-        reaction.track(() => {
-            rendering = baseComponent(props as P)
+        let rendering!: React.ReactElement<any> | null
+        observerReaction.track(() => {
+            rendering = baseComponent(props)
         })
         return rendering
     })
     memoComponent.displayName = baseComponentName
-    return memoComponent
+    return memoComponent as any
 }
 
 function useForceUpdate() {
     const [tick, setTick] = useState(1)
-    return () => {
+
+    const update = useCallback(() => {
         setTick(tick + 1)
-    }
+    }, [])
+
+    return update
 }
 
-function useUnmount(fn) {
-    useEffect(() => fn, [])
+function useObserverReaction(baseComponentName: string) {
+    // forceUpdate 2.0
+    const forceUpdate = useForceUpdate()
+
+    const lastReaction = useRef<Reaction | undefined>(undefined)
+    useUnmount(() => {
+        if (lastReaction.current) {
+            lastReaction.current.dispose()
+            lastReaction.current = undefined
+        }
+    })
+
+    // we create a reaction per render to make sure we only re-render when observed values change OUTSIDE the render phase
+    if (lastReaction.current) {
+        lastReaction.current.dispose()
+        lastReaction.current = undefined
+    }
+
+    lastReaction.current = new Reaction(`observer(${baseComponentName})`, forceUpdate)
```

Comment:
It was a fix for this:
Imagine the case of useObservableProps for example and computed values based on those observable props

1. observableProps are updated on the beginning of the render, but that's ok since they are untracked with an action
2. computed values are updated at this point as well, and since they are being tracked by the old track function they call to set state of force update 
3. the remainder of the render function runs as usual, using the new computed values
4. here is the problem, the setState done inside forceUpdate was previously scheduled, so it will render once more with the new computed values (which were previously used on the last render anyway)

With this change only changes detected outside the render phase will actually call force update

As for memory leaks, it is disposed before each render and on unmount, so there should be no memory leaks AFAIK

As for performance it is true it might be slower, but I'm not sure by how much. Makes me wonder, is there a way to "reset" what is tracked by a Reaction? If so then it could be memoized and just call this "reset" before each render phase is triggered and it would achieve the same effect.

Reply:
Well, I would not rather slow down universal `observer` in an attempt to optimize a single (possibly rarely used) hook.

The point still remains, the observables should be created once and its reference should not change afterward just because underlying values have changed. That beats all the purpose.

It's totally possible I am missing something here. I don't really understand all the intricacies of MobX. I am pretty much just a consumer and I just grabbed a code from @mweststrate and made it into a package :) I can imagine that he memoized the reaction with a good reason.

Regarding memory leak, I don't think it would work. The `useObservableEffect` would throw away reaction only if user specified inputs change, not when the observables are changed. And even if it would work the way you are saying, once again it's so much wasteful to throw a bunch of stuff away on each render and reinitialize it.

Reply:
Reverted to memoized version and changed tests to accept double renderings in cases where computeds are based on observable props